### ISSUE TITLE
WebSocket sporadic ReadPendingException using suspend/resume

### ISSUE THREAD

**Issue by robertvazan** [eclipse/jetty.project#3968]
I am sporadically getting the following warning followed by an exception in production:

```
WARN org.eclipse.jetty.io.FillInterest - Read pending for AC.ReadCB@2d603ff3{WebSocketServerConnection@2d603ff3::SocketChannelEndPoint@7716547f{/127.0.0.1:34832<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=0/1200000}{io=1/1,kio=1,kro=1}->WebSocketServerConnection@2d603ff3[s=ConnectionState@4f677843[OPENED],f=org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection$Flusher@31e71450[IDLE][queueSize=0,aggregateSize=-1,terminated=null],g=Generator[SERVER,validating],p=Parser@3744d188[ExtensionStack,s=START,c=0,len=51,f=null]]} prevented AC.ReadCB@2d603ff3{WebSocketServerConnection@2d603ff3::SocketChannelEndPoint@7716547f{/127.0.0.1:34832<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=0/1200000}{io=1/1,kio=1,kro=1}->WebSocketServerConnection@2d603ff3[s=ConnectionState@4f677843[OPENED],f=org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection$Flusher@31e71450[IDLE][queueSize=0,aggregateSize=-1,terminated=null],g=Generator[SERVER,validating],p=Parser@3744d188[ExtensionStack,s=START,c=0,len=51,f=null]]}
[VHostServer-16] WARN org.eclipse.jetty.util.thread.strategy.EatWhatYouKill -
java.nio.channels.ReadPendingException
	at org.eclipse.jetty.io.FillInterest.register(FillInterest.java:58)
	at org.eclipse.jetty.io.AbstractEndPoint.fillInterested(AbstractEndPoint.java:361)
	at org.eclipse.jetty.io.AbstractConnection.fillInterested(AbstractConnection.java:134)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.fillInterested(AbstractWebSocketConnection.java:336)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:464)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:428)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:781)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:917)
	at java.lang.Thread.run(Thread.java:748) 
```

Nothing before or after that. Version 9.4.19.v20190610. There was no such exception in 9.4.10.v20180503.

Not sure whether this is related to #1727 or #1047.

**Comment by robertvazan** [comment]
I am also getting this exception (occasionally) when calling `SuspendToken.resume()`:

```
WARN org.eclipse.jetty.io.FillInterest - Read pending for AC.ReadCB@2128350d{WebSocketServerConnection@2128350d::SocketChannelEndPoint@30b50c87{/127.0.0.1:43026<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=13/1200000}{io=1/1,kio=1,kro=1}->WebSocketServerConnection@2128350d[s=ConnectionState@42cfffa7[OPENED],f=org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection$Flusher@58e810e6[IDLE][queueSize=0,aggregateSize=-1,terminated=null],g=Generator[SERVER,validating],p=Parser@7a64c27c[ExtensionStack,s=START,c=0,len=57,f=null]]} prevented AC.ReadCB@2128350d{WebSocketServerConnection@2128350d::SocketChannelEndPoint@30b50c87{/127.0.0.1:43026<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=14/1200000}{io=1/1,kio=1,kro=1}->WebSocketServerConnection@2128350d[s=ConnectionState@42cfffa7[OPENED],f=org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection$Flusher@58e810e6[IDLE][queueSize=0,aggregateSize=-1,terminated=null],g=Generator[SERVER,validating],p=Parser@7a64c27c[ExtensionStack,s=START,c=0,len=57,f=null]]}
ERROR myapp.MyClass - Caught exception
java.nio.channels.ReadPendingException
	at org.eclipse.jetty.io.FillInterest.register(FillInterest.java:58)
	at org.eclipse.jetty.io.AbstractEndPoint.fillInterested(AbstractEndPoint.java:361)
	at org.eclipse.jetty.io.AbstractConnection.fillInterested(AbstractConnection.java:134)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.fillInterested(AbstractWebSocketConnection.java:336)
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.resume(AbstractWebSocketConnection.java:620)
	(... application call stack ...)
```

I am wondering whether 9.4.19.v20190610 is stable enough for use.

**Comment by joakime** [comment]
How are you using suspend / resume?
We have many organizations using it now without the problems you are reporting.

Perhaps there's something about how you use it that is aggravating these exceptions?
Please be as detailed as you can, especially about which thread is doing the suspend, and which is doing the resume.

**Comment by robertvazan** [comment]
It has been some time since I wrote the code, but here's an overview of what it does. I implemented jetty websocket interfaces / base classes:

```java
class ApiListener extends WebSocketServlet {
	@Override public void configure(WebSocketServletFactory factory) {
		// (some WebSocketServletFactory setup here)
		factory.setCreator(new ApiCreator());
	}
}
class ApiCreator implements WebSocketCreator {
	@Override public Object createWebSocket(ServletUpgradeRequest request, ServletUpgradeResponse response) {
		// (some ServletUpgradeRequest checks here)
		return new ApiSocket(/* ... */);
	}
}
class ApiSocket implements WebSocketListener {
	// ...
}
```

`ApiSocket` (`WebSocketListener`) then overrides `on*` methods:

```java
@Override public synchronized void onWebSocketConnect(Session session) {
	this.session = session;
	// (some application setup here, no communication)
}
@Override public synchronized void onWebSocketBinary(byte[] payload, int offset, int length) {
	byte[] data = Arrays.copyOfRange(payload, offset, offset + length);
	CompletableFuture<?> future = process(data);
	if (!future.isDone()) {
		++backpressure;
		if (suspension == null)
			suspension = session.suspend();
		future.whenCompleteAsync((r, e) -> {
			synchronized (this) {
				--backpressure;
				if (e != null)
					fail();
				else if (suspension != null && backpressure <= 0) {
					SuspendToken resumed = suspension;
					suspension = null;
					resumed.resume();
				}
			}
		}, executor);
	} else
		future.join();
}
@Override public synchronized void onWebSocketText(String message) {
	// (something similar to onWebSocketBinary here)
}
@Override public synchronized void onWebSocketError(Throwable cause) {
	fail();
}
@Override public synchronized void onWebSocketClose(int code, String reason) {
	if (code == StatusCode.NORMAL) {
		// (some app logic here, already started writes may continue for some time)
	} else
		fail();
}
private void fail() {
	// (some logic here that causes writes to cease)
	session.disconnect();
}
```

The suspension logic is a bit complicated, but it is intended to work correctly even if (perhaps already buffered) data arrives while the socket is suspended.

I have omitted exception handling for brevity. Basically, every method as well as the `CompletableFuture` callback is wrapped in a try-catch that calls `fail()`. Method `fail()` never throws - it swallows all exceptions.

This is just socket reading. Parallel to this, socket writes continue in the background since the socket is opened until it is closed. Socket writes are done via the asynchronous version of `sendBytes` / `sendString` that takes a callback. Only one write runs at a time.

**Comment by joakime** [comment]
Those `synchronized` methods have a strong code smell to me.
But I would need to test that assumption.

**Comment by robertvazan** [comment]
None of the methods block, so synchronization should just improve reliability with no undesirable side effects. All writes are scheduled on a thread pool, so there's no risk of recursive read->write->read calls. Also, as I am looking at the code, I realize that even `session.close()` is always scheduled on a thread pool. Another thing I omitted is that `ApiSocket` maintains a state variable and checks it in every method to ensure the methods aren't called in funny order by either jetty (for reads) or the rest of the application (for writes).

**Comment by lachlan-roberts** [comment]
@robertvazan I have found a race condition introduced with the changes to WebSocket suspend in 9.4.19 (#3498) which looks like it is the cause of this issue and of issue #3967

I would not use 9.4.19 which also has a WebSocket memory leak #3871 fixed in 9.4.20, but I would suggest using 9.4.18 until 9.4.21 comes where both issues should be fixed

**Comment by robertvazan** [comment]
@lachlan-roberts Thanks for the info. I am reverting to older version for now.
